Userland/Libraries/LibWasm/Parser/Parser.cpp

```diff
@@ -0,0 +1,1323 @@
+        size_t count;
+        if (!LEB128::read_unsigned(stream, count))
+            return ParseResult<Vector<ResultT>> { with_eof_check(stream, ParseError::ExpectedSize) };
+
+        Vector<ResultT> entries;
+        for (size_t i = 0; i < count; ++i) {
+            auto result = T::parse(stream);
+            if (result.is_error())
+                return ParseResult<Vector<ResultT>> { result.error() };
+            entries.append(result.release_value());
+        }
+        return ParseResult<Vector<ResultT>> { move(entries) };
+    } else {
+        size_t count;
+        if (!LEB128::read_unsigned(stream, count))
+            return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::ExpectedSize) };
+
+        Vector<T> entries;
+        for (size_t i = 0; i < count; ++i) {
+            if constexpr (IsSame<T, size_t>) {
+                size_t value;
+                if (!LEB128::read_unsigned(stream, value))
+                    return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::ExpectedSize) };
+                entries.append(value);
+            } else if constexpr (IsSame<T, ssize_t>) {
+                ssize_t value;
+                if (!LEB128::read_signed(stream, value))
+                    return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::ExpectedSize) };
+                entries.append(value);
+            } else if constexpr (IsSame<T, u8>) {
+                if (count > 64 * KiB)
+                    return ParseResult<Vector<T>> { ParseError::HugeAllocationRequested };
+                entries.resize(count);
+                if (!stream.read_or_error({ entries.data(), entries.size() }))
+                    return ParseResult<Vector<T>> { with_eof_check(stream, ParseError::InvalidInput) };
+                break; // Note: We read this all in one go!
+            }
+        }
+        return ParseResult<Vector<T>> { move(entries) };
+    }
+}
+
+static ParseResult<String> parse_name(InputStream& stream)
+{
+#if WASM_BINPARSER_DEBUG
+    ScopeLogger logger;
+#endif
+    auto data = parse_vector<u8>(stream);
+    if (data.is_error())
+        return data.error();
+
+    return String::copy(data.value());
+}
+
+template<typename T>
+struct ParseUntilAnyOfResult {
+    u8 terminator { 0 };
+    Vector<T> values;
+};
+template<typename T, typename... Args>
+static ParseResult<ParseUntilAnyOfResult<T>> parse_until_any_of(InputStream& stream, Args... terminators)
+{
+#if WASM_BINPARSER_DEBUG
+    ScopeLogger logger;
+#endif
+    ReconsumableStream new_stream { stream };
+    ScopeGuard drain_errors {
+        [&] {
+            new_stream.handle_any_error();
+        }
+    };
+
+    ParseUntilAnyOfResult<T> result;
+    for (;;) {
+        u8 byte;
+        new_stream >> byte;
+        if (new_stream.has_any_error())
+            return with_eof_check(stream, ParseError::ExpectedValueOrTerminator);
+
+        if ((... || (byte == terminators))) {
```

Comment:
This is... a left fold across the terminators? I had no idea you could do it that way. 

That being said, variadic args for this use case seems a bit excessive? I probably would have used a Span\<const u8\> myself 🤔 .

Reply:
But why not do it at compiletime if we can?
(Can I even make a ReadonlyBytes with a bunch of literals? I'd have to put them in an array outside the function call...)

Reply:
Ah you know, for some  reason I was thinking you could construct ReadonlyBytes from a `std::initailizer_list<int>` or similar, like passing ` parse_until_error<Instruction>(blah, { 0x09, 0x05 });` but that won't work at all. :)

Only technical reason would be "binary size" but not like that's super important right now.

Reply:
I could lower it one level and make it take a `const Array<u8, N>&`.
With the current usage there will still be two copies of this function though (one for a single-element array, and one for a two-element array)

Though do note that there are only two terminators in the spec right now.

Reply:
After I left that comment I took a peek at your original MR and saw that you promoted the terminators to template arguments anyway, so not sure if that would be a PITA to rebase the next set of commits on top of :). There's only 3 uses of it anyway, so not super high priority.

Reply:
Then let's keep it as-is in this PR, and I'll change the function to take an array in the next set of commits (I'll check just how much the binary size changes, maybe it won't be particularly significant 🤷 )